bpf/bpf_overlay.c

```diff
@@ -323,6 +329,49 @@ int tail_handle_ipv4(struct __ctx_buff *ctx)
 					      CTX_ACT_DROP, METRIC_INGRESS);
 	return ret;
 }
+
+#ifdef ENABLE_VTEP
+/*
+ * ARP responder for ARP requests from VTEP
+ * Respond to remote VTEP endpoint with cilium_vxlan MAC
+ */
+__section_tail(CILIUM_MAP_CALLS, CILIUM_CALL_ARP)
+int tail_handle_arp(struct __ctx_buff *ctx)
+{
+	union macaddr mac = NODE_MAC;
+	union macaddr smac;
+	__be32 sip;
+	__be32 tip;
+	__u32 monitor = 0;
+	int i;
+	struct remote_endpoint_info *info;
+	struct bpf_tunnel_key key = {};
+
+	if (unlikely(ctx_get_tunnel_key(ctx, &key, sizeof(key), 0) < 0))
+		return DROP_NO_TUNNEL_KEY;
+	if (key.tunnel_id != WORLD_ID)
+		return DROP_INVALID_VNI;
+
+	if (!arp_validate(ctx, &mac, &smac, &sip, &tip))
+		return CTX_ACT_DROP;
+	if (!__lookup_ip4_endpoint(tip))
+		return CTX_ACT_DROP;
+	if (arp_prepare_response(ctx, &mac, tip, &smac, sip) < 0)
+		return CTX_ACT_DROP;
+
+	info = lookup_ip4_remote_endpoint(sip);
+	if (!info)
+		return DROP_NO_TUNNEL_ENDPOINT;
+	for (i = 0; i < VTEP_NUMS; i++) {
+		if (info->tunnel_endpoint == VTEP_ENDPOINT[i])
+			return __encap_and_redirect_with_nodeid(ctx, info->tunnel_endpoint,
+									info->sec_label, monitor);
+	}
+
+	return CTX_ACT_OK;
```

Comment:
I don't think it makes sense to prepare the ARP response and then pass that ARP response to the stack (eg if there is no configured VTEP for that tunnel endpoint). What will the stack do with that packet?

Reply:
yep, I have not thought  about what the last `return CTX_ACT_OK;` would do, I can remove it.

Reply:
Cilium generally speaking tries to integrate with the stack, which is to say that we figure out whether Cilium must control something (eg ARP for IP addresses that Cilium manages), then for that traffic specifically, apply some change (eg reply to ARP). This way, if it turns out that the user still needs some other Linux functionality or they want to configure some other part of the stack to handle traffic that Cilium shouldn't handle, it will "just work".

The only exception to the above rule is where maybe for security or architectural reasons, Cilium should block or specially handle the traffic.

Given that today, Cilium just does `CTX_ACT_OK` for this path (based on the `from_overlay()` code below where this PR edits), I think it likely makes sense to follow this same approach there. ie Cilium shouldn't just randomly drop ARP requests just because it doesn't know enough information to handle it. That is to say, I expect this `CTX_ACT_OK` case to still be here for any traffic that we don't explicitly _need_ to handle for this feature to function correctly.

Reply:
ok, I will keep the last `CTX_ACT_OK`

I also looked at `arp_respond` in the `bpf/lib/arp.h` , I wonder if I should also return `return send_drop_notify_error(ctx, 0, ret, CTX_ACT_DROP, METRIC_EGRESS);` when `arp_prepare_response(ctx, &mac, tip, &smac, sip) !=0` instead of  `return CTX_ACT_DROP;` or return `CTX_ACT_OK`  here too ?

``` 
arp_respond(struct __ctx_buff *ctx, union macaddr *smac, __be32 sip,
            union macaddr *dmac, __be32 tip, int direction)
{
        int ret = arp_prepare_response(ctx, smac, sip, dmac, tip);

        if (unlikely(ret != 0))
                goto error;

        cilium_dbg_capture(ctx, DBG_CAPTURE_DELIVERY,
                           ctx_get_ifindex(ctx));
        return ctx_redirect(ctx, ctx_get_ifindex(ctx), direction);

error:  
        return send_drop_notify_error(ctx, 0, ret, CTX_ACT_DROP, METRIC_EGRESS);
}

Reply:
so base on letting unknown packet through idea, I think I should also return `CTX_ACT_OK` when  `(!info)` ?

Reply:
return `CTX_ACT_OK`

Reply:
I think it makes sense to send_drop_notify_error() on failure here, yes.

Regarding this last section, I think I failed to communicate what I was worried about with this specific later part of the function. At this point, the packet has been changed from an ARP request into an ARP response. Once Cilium has modified the packet like this, it is now acting as an ARP responder and should either successfully hand the response back towards the request, or fail out / drop it. The final `CTX_ACT_OK` here is instead changing the request from some remote note (or VTEP) into a response, and then passing the response to the stack. This makes no sense, as the stack never made that request in the first place.